packages/flutter/lib/src/foundation/tree_diagnostics_mixin.dart

```diff
@@ -0,0 +1,38 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+/// A mixin that helps dump string representations of trees.
+class TreeDiagnosticsMixin {
+  @override
+  String toString() => '$runtimeType#$hashCode';
+
+  /// Returns a string representation of this node and its descendants.
+  String toStringDeep([String prefixLineOne = '', String prefixOtherLines = '']) {
```

Comment:
we should have a test that verifies the output here is sane (just a dummy TreeDiagnosticsMixin class with an object hierarchy with a root with two children and one of those children has a child, and then compare the output to a known-good string).